### PR TITLE
Static timing implementation

### LINKED ISSUE #153: Static timing implementation proposal
Add statically timed groups to Futil. There are five **Q** marked in this proposal. If you are @sampsyo or @sgpthomas, please address these in comments.

Note that this proposal does attempt to implement [SDC scheduling](https://www.csl.cornell.edu/~zhiruz/pdfs/sdc-dac2006.pdf). See the discussion at the end.

### Futil syntax
Add the following syntax to groups:
```
group foo<k1=n1, k2=n2> { ... }
```
where `k` is a double-quoted strings and `n` is a decimal number.

For static timing implementation, we will add information using the string `"static"`:
```
group foo<"static"=3> { ... } // 3 cycle delay between go and done signals
```

The rationale for having a general key-value syntax instead of a specific one is to allow addition of information in the future. The downside of this choice is that it is not clear that the relationship described is between the timing of the `go` and the `done` hole.

### Library syntax
The current syntax for a primitive definition uses the syntax:
```
primitive std_reg[p_width](in: p_width) -> (out: p_width, done: 1) 
```

We will extend this to:
```
primitive std_reg<k1=n1, k2=n2>[p_width](in: p_width) -> (out: p_width, done: 1) 
```

**Q1** (@sampsyo the syntax is getting ugly. Thoughts?)

## Adding timing information
In traditional HLS, static timing information is associated with the primitive operations such as `+` and `*`. In our case, we will extend the primitive library to provide information about the static timing of various primitives. 

For example, the current standard register definition will be changed to:
```
primitive std_reg<"static"=1>[p_width](...) -> (...)
```
which denotes that the done signal is raised one cycle after the current cycle.

### Combinational primitives (static=0)
A funky corner case will the implementation is that combinational circuits have a 0 delay -- the `done` signal is raised in the same cycle as the `go` (in the actual implementation, there doesn't seem to be any `done` signal provided to the combinations primitives. Not sure why. @sgpthomas would know.).

A few considerations:
1. We need to allow optimizations like operator chaining to be implemented. This in turn requires is to explicitly talk about the timing of combinational circuits and when to register their values.
2. If we add an explicit `done` signal to combinational primitives, we'll be carrying an extra wire (the section on **Done elimination** might get rid of these wires)

The tentative proposal is to add `"static=0"` to the primitives and not do anything interesting with it.

**Q2** Open to comments/ideas here

## Statically timed FSM generation pass

The passes uses the information from the previous stages to:
- Generate optimized FSMs when the groups in a control are statically timed.
- Propagate timing information to newly generated groups from compiling nested control.
- Remove `done` holes from groups that don't need it.

The pass will run before the default `compile_control` pass and rely on it to generate valid, inlinable programs for groups that cannot be statically timed.

### Generating optimized FSMs
The following shows the differences in the code generated for `seq { one; two; three; }` when all sub-groups have timing information.

```diff
cells {
   ...
   fsm = std_reg(32); // register to track FSM state
}
-seq0 {
+seq0<"static" = n> {
  // new compilation group. For brevity, assume all
  // assigns are guarded by seq0[go]

  // FSM state updates
- fsm.in = fsm.out == 0 & one[done] ? 1;
- fsm.in = fsm.out == 1 & two[done] ? 2;
- fsm.in = fsm.out == 2 & three[done] ? 3;
+ fsm.in = fsm.out != 3 ? fsm.in + 1; // fake syntax, needs to be a port.

  // Group go signals
- one[go] = fsm.out == 0 & !one[done];
+ one[go] = fsm.out == 0 ;
- two[go] = fsm.out == 1 & !two[done];
+ two[go] = fsm.out == 1;
- three[go] = fsm.out == 2 & !two[done];
+ three[go] = fsm.out == 2;

  // Group is done
  seq0[done] = fsm.out == 3;
}
```

**Q3** Is it actually correct to not read the `done` signals from the subgroups? Need hard thinking.

### Notes
1. The initial implementation *will not* allow partially statically timed groups -- all sub-groups must be statically timed for the pass to generate optimized FSMs. This restriction should be lifted in a future PR to allow for partially timed.
2. When compiling `while` groups, the generated group will not have any timing information associated to it. This will allow the default compile control compile it like any other group.

### Removing `done` holes
**Q4** Primitives like `std_reg` actually has two internal registers -- one to store the data and one to store the done signal. A pass to remove `done` holes by rewriting to a simpler `std_reg_no_done` primitive might help us to get rid of this extra overhead.

## SDC Scheduling
**Q5** SDC scheduling is about many things -- parallelizing dataflow graphs, reducing slack for each stage, chaining operators etc. An open question is how we should handle this.

### PR DESCRIPTION
**Pull request by rachitnigam**
Fixes #153.

Implements the `StaticTiming` pass that compiles groups using static FSMs when
possible and calculates the static of a group.

This PR implements
- The static timing proposal
  - Adds attribute syntax `group g<"k"=n, ...> { ... }`
  - Implements a pass that uses the `static` attribute to generate groups that can use self incrementing counters to enable and disable `go`/`done` holes.
- A pass manager
  - Adds the `register_pass!` and `register_alias!` macros.
  - Adds the `-d` flag to the CLI to disable a pass or an alias.
- New macros for writing structure components.
  - `guard!`, and `add_wires!`.
  - Renames `add_structure!` to `structure!`.
  - Removes `comp."port"` syntax for `port!`. Only have `comp["out"]` for consistency.
- Change the inliner to remove all writes to holes even if there are no reads from them. This was needed because `done` holes are not read from static groups.

### CODE REVIEW

File: calyx/src/backend/verilog/gen.rs [hunk 1]
```diff
@@ -53,7 +53,8 @@ fn validate_guard(guard: &GuardExpr) -> bool {
         }
         GuardExpr::Not(inner) => validate_guard(inner),
         GuardExpr::Atom(Atom::Port(p)) => {
-            matches!(p, Port::Comp { .. } | Port::This { .. })
+            let ret = matches!(p, Port::Comp { .. } | Port::This { .. });
```
Comment (rachitnigam) [inline calyx/src/backend/verilog/gen.rs:56]: Unnecessary change

File: calyx/src/backend/verilog/gen.rs [hunk 2]
```diff
@@ -326,9 +327,10 @@ fn wire_id_from_port<'a>(port: &Port) -> D<'a> {
         Port::Comp { component, port } => {
             D::text(format!("{}_{}", component.to_string(), port.to_string()))
         }
-        Port::Hole { .. } => unreachable!(
-            "This should have been caught in the validation checking"
-        ),
+        Port::Hole { group, name } => unreachable!(format!(
+            "Structure has a group hole: {}[{}]",
+            group.id, name.id
+        )),
```
Comment (rachitnigam) [inline calyx/src/backend/verilog/gen.rs:333]: Note for @sgpthomas: It is not caught by the validator for some reason. Remind me to write a test case.

**Comment by rachitnigam** [comment]
TODO:
- [ ] Tests for if compilation (manual tests work)
- [ ] Compile while groups with static information.

File: calyx/src/backend/verilog/gen.rs (hunk 2 above)
Comment (rachitnigam) [reply calyx/src/backend/verilog/gen.rs:333]: Specifically, I made the changes to the inliner that @sgpthomas needs to review.

Comment (sgpthomas) [reply calyx/src/backend/verilog/gen.rs:333]: where are the changes to the inliner? are they in this branch?

Comment (rachitnigam) [reply calyx/src/backend/verilog/gen.rs:333]: Yeah, they should be. Specifically the inlined wouldn’t delete holes if there were no reads from them.

File: calyx/src/passes/inliner.rs [hunk 3]
```diff
@@ -118,35 +98,40 @@ fn inline_hole(st: &mut StructureGraph, hole: String) {
...
     }
 
     // flatten groups (maybe temporary)
     for idx in st.edge_idx().detach() {
         st.get_edge_mut(idx).group = None;
     }
     st.groups = HashMap::new();
-    st.groups.insert(None, st.edge_idx().collect());
+    st.groups
+        .insert(None, (HashMap::new(), st.edge_idx().collect()));
```
Comment (sgpthomas) [inline calyx/src/passes/inliner.rs:134]: did you add a test to make sure that this is correct? It should be but I remember I did the `edges_inlined` thing for a particular reason. Ofc I don't remember what it was exactly. Curse myself for not writing this down then...

Comment (rachitnigam) [reply calyx/src/passes/inliner.rs:134]: I can try adding one.
